File: packages/cexplorer-api/src/endpoints/article.ts
import type { ArticleDetailResponse, ArticleListResponse } from "@/types/articleTypes";

import { handleFetch } from "@/lib/handleFetch";

interface ArticleDetailProps {
  lng: "en";
  type: "page" | "article";
  url: string;
}

/**
 * Fetches the detail of a specific article from the Cexplorer API.
 *
 * Requires language, type, and article URL as parameters to identify the article.
 *
 * @param {ArticleDetailProps} params - Object containing article parameters.
 * @param {string} params.lng - The language code (e.g. "en", "cs").
 * @param {string} params.type - The article type (e.g. "news", "blog").
 * @param {string} params.url - The unique URL slug of the article.
 *
 * @example
 * ```ts
 * const detail = await getArticleDetail({ lng: "en", type: "news", url: "cardano-updates" });
 * ```
 *
 * @returns {Promise<ArticleDetailResponse>} A promise resolving to the article detail response.
 * @throws If the request fails or the article parameters are invalid.
 */
export const getArticleDetail = async ({ lng, type, url }: ArticleDetailProps) => {
  const apiUrl = "/article/detail";

  const options = {
    params: {
      lng,
      type,
      url,
    },
  };

  return handleFetch<ArticleDetailResponse>(apiUrl, undefined, options);
};

/**
 * Fetches a paginated list of articles from the Cexplorer API.
 *
 * This function retrieves articles in the specified language with optional filtering
 * by category. The result is paginated using `limit` and `offset`.
 *
 * @param {Object} params - Parameters for fetching articles.
 * @param {"en"} params.lng - Language code (currently only "en" is supported).
 * @param {number} params.offset - Offset for pagination.
 * @param {number} params.limit - Number of articles to fetch.
 * @param {string} [params.category] - Optional category filter.
 *
 * @example
 * ```ts
 * const articles = await getArticleList({ lng: "en", offset: 0, limit: 10 });
 * ```
 *
 * @returns {Promise<ArticleListResponse>} A promise resolving to the list of articles.
 * @throws If the request fails or parameters are invalid.
 */
export const getArticleList = async ({ lng, offset, limit, category }: { lng: "en"; offset: number; limit: number; category?: string }) => {
  const apiUrl = "/article/list";

  const options = {
    params: {
      lng,
      type: "article",
      limit,
      offset,
      category,
    },
  };

  return handleFetch<ArticleListResponse>(apiUrl, offset, options);
};

File: packages/cexplorer-api-rs/src/endpoints/article.rs
use crate::client::fetch_with_params;
use crate::error::CexplorerError;
use crate::types::article_types::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct ArticleDetailParams {
    pub lng: String,
    #[serde(rename = "type")]
    pub article_type: String,
    pub url: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ArticleListParams {
    pub lng: String,
    #[serde(rename = "type")]
    pub article_type: String,
    pub limit: u64,
    pub offset: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
}

pub async fn get_article_detail(
    lng: &str,
    article_type: &str,
    url: &str,
) -> Result<ArticleDetailResponse, CexplorerError> {
    let endpoint = "/article/detail";
    let params = ArticleDetailParams {
        lng: lng.to_string(),
        article_type: article_type.to_string(),
        url: url.to_string(),
    };
    fetch_with_params::<ArticleDetailResponse, ArticleDetailParams>(endpoint, Some(&params)).await
}

pub async fn get_article_list(
    lng: &str,
    offset: u64,
    limit: u64,
    category: Option<&str>,
) -> Result<ArticleListResponse, CexplorerError> {
    let endpoint = "/article/list";
    let params = ArticleListParams {
        lng: lng.to_string(),
        article_type: "article".to_string(),
        limit,
        offset,
        category: category.map(|s| s.to_string()),
    };
    fetch_with_params::<ArticleListResponse, ArticleListParams>(endpoint, Some(&params)).await
}
